core/src/events/mod.rs

```diff
@@ -0,0 +1,256 @@
+use alloc::sync::Arc;
+use codec::{Decode, Encode};
+use scale_decode::{DecodeAsFields, DecodeAsType};
+use subxt_metadata::PalletMetadata;
+
+use crate::{Config, Error, Metadata, MetadataError};
+
+use alloc::vec::Vec;
+
+/// Trait to uniquely identify the events's identity from the runtime metadata.
+///
+/// Generated API structures that represent an event implement this trait.
+///
+/// The trait is utilized to decode emitted events from a block, via obtaining the
+/// form of the `Event` from the metadata.
+pub trait StaticEvent: DecodeAsFields {
+    /// Pallet name.
+    const PALLET: &'static str;
+    /// Event name.
+    const EVENT: &'static str;
+
+    /// Returns true if the given pallet and event names match this event.
+    fn is_event(pallet: &str, event: &str) -> bool {
+        Self::PALLET == pallet && Self::EVENT == event
+    }
+}
+
+/// A phase of a block's execution.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Decode, Encode)]
+pub enum Phase {
+    /// Applying an extrinsic.
+    ApplyExtrinsic(u32),
+    /// Finalizing the block.
+    Finalization,
+    /// Initializing the block.
+    Initialization,
+}
+
+/// The event details.
+#[derive(Debug, Clone)]
+pub struct EventDetails<T: Config> {
```

Comment:
How hard would it be to port over `Events` too? 

Right now, the "entry point" for working with events in `subxt_core` is to get hold of a single event from some events bytes and then call `Events::decode_from` to decode them.

It'd be much nicer I think if the entry point was to get bytes representing all of the events (which is how they come back from a node) and then calling `Events::decode_from` on the bytes. To make this work I reckon we could:

- Move `Events` over to `subxt_core`, too
- Rename `Events::new` to `Events::decode_from` and make it pub.
- Get rid of `Events::block_hash()` and `ExtrinsicEvents::block_hash()`: it looks like this is just here to be helpful but isn't actually needed internally, and if people want to know the block hash then they can fetch the block first and then ask for events from it etc. This allows us to get rid of `block_hash` from `Events::decode_from`, making it cleaner.
- Make `Events::new_from_client` a freestanding function exposed in `subxt::events::new_from_client`; now we don't have anything mentioning `OnlineClient` in Events I think :)